Improve listing resources across clusters

rosstimothy opened the pull request:
Previously doing an `ls -R` would list all resources per profile in parallel. This had little benefit when one profile needed to list multiple clusters as those were still done serially. To mitigate this this things are broken up into two separate steps:

1) Connect to all clusters the user has credentials for in parallel 
2) List resources for each cluster in parallel

This ensures that by the time a client is established for all available clusters that resource listing can happen in parallel regardless of profile. All operations are still restricted to an upper bound, however those limits wer also bumped to help performance.

ravicious commented on tool/tsh/db.go:
```diff
@@ -135,59 +135,57 @@ func (l databaseListings) Swap(i, j int) {
+			for _, database := range databases {
+				localDBListings = append(localDBListings, databaseListing{
+					Proxy:    cluster.profile.ProxyURL.Host,
+					Cluster:  cluster.tc.SiteName,
```
Should this use `cluster.name` instead?

ravicious commented on tool/tsh/tsh.go:
```diff
@@ -1898,6 +1899,63 @@ func onListNodes(cf *CLIConf) error {
+// clusterClient is a client for a particular cluster
+type clusterClient struct {
+	name    string
+	tc      *client.TeleportClient
```
Nit: I'm not an expert here but exposing `TeleportClient` within `clusterClient` is slightly risky to me. `TeleportClient` is not safe for concurrent use while `clusterClient` was made specifically for use in a concurrent context. We could get away with exposing just the return value of `tc.DefaultResourceFilter`, couldn't we?

ravicious commented on tool/tsh/db.go:
```diff
@@ -135,59 +135,57 @@ func (l databaseListings) Swap(i, j int) {
 }
 
 func listDatabasesAllClusters(cf *CLIConf) error {
```
`tsh db ls -R` doesn't work for me with these changes. I tried `tsh db ls -Rv` but it doesn't print anything either.

ravicious commented on tool/tsh/tsh.go:
```diff
@@ -1926,23 +1926,32 @@ func getClusterClients(cf *CLIConf) ([]*clusterClient, error) {
+			// log error and create a site for the proxy we were able to
+			// connect to so that results are still retrieved.
+			logger.Errorf("Failed to lookup leaf clusters: %v", err)
+			sites = []types.Site{{Name: profile.Cluster}}
```
What do you think about always logging these errors to stderr?

Right now they'll appear only if the user also provides the `-d` flag, otherwise the error will be silent. Assuming a cluster is down and I don't know how many resources it has (or if any), if I do `tsh ls -R` and I won't see any nodes from that cluster, I'll treat the silent failure as success and assume that the cluster doesn't have any nodes.

ravicious replied:
Yeah, this is pretty nice!

rosstimothy committed:
Improve listing resources across clusters

Previously doing an `ls -R` would list all resources per profile in
parallel. This had little benefit when one profile needed to list
multiple clusters as those were still done serially. To mitigate this
things are broken up into two separate steps:

  1) Connect to all clusters the user has credentials for in parallel
  2) List resources for each cluster in parallel

This ensures that by the time a client is established for all available
clusters that resource listing can happen in parallel regardless of
profile. All operations are still restricted to an upper bound, however
those limits were also bumped to help performance.

rosstimothy committed:
Merge branch 'master' into tross/tsh_ls_r

rosstimothy committed:
Merge branch 'master' into tross/tsh_ls_r
